rfcs/0042-omnilock/0042-omnilock.md

```diff
@@ -0,0 +1,558 @@
+option IdentityOpt (Identity);
+
+// the data structure used in lock field of witness
+table OmniLockWitnessLock {
+    signature: BytesOpt,
+    omni_identity: IdentityOpt,
+    preimage: BytesOpt,
+}
+```
+
+When `omni_identity` is present, it will be validated whether the provided auth in `omni_identity` is present in RC
+AdminList Cell associated with the current lock script via SMT validation rules. In this case, the auth included in
+`omni_identity` will be used in further validation.
+
+If `omni_identity` is missing, the auth included in lock script args will then be used in further validation.
+
+Once the processing above is successfully done and the auth to be used is confirmed, the flag in the designated auth
+will be checked for the succeeding operations:
+
+* When the auth flag is 0x0, a signature must be present in `OmniLockWitnessLock`. We will use the signature for secp256k1
+  recoverable signature verification. The recovered public key hash using the blake160 algorithm must match the current
+  auth content.
+
+* When the auth flag is 0xFC, we will check against the current transaction, and there must be an input cell, whose lock
+  script matches the auth content when hashed via blake160.
+
+
+When `signature` is present, the signature can be used to unlock the cell in anyone-can-pay mode.
+
+When `preimage` is present, the auth flag can be 0xFD or 0xFE.
+
+* When auth flag is 0xFD, the `exec` method is used. The preimage's memory layout will be as follows:
+```
+exec code hash (32 bytes)
+exec hash type (1 byte)
+place (1 byte)
+bounds (8 bytes)
+pubkey blake160 hash (20 bytes)
+```
+
+The `place` and `bounds` are passed in directly to [exec](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec):
+```C
+int ckb_exec(size_t index, size_t source, size_t place, size_t bounds, int argc, char* argv[]);
+```
+The `index` is located by `exec code hash` and `exec hash type` from `cell_deps`([See usage](https://github.com/nervosnetwork/ckb-c-stdlib/blob/20578dfb092b3b3761df755395e20ec142a83d6e/ckb_syscalls.h#L368)].
+Finally, message, signature, pubkey blake160 hash are encoded into hex strings. Then these hex strings are passed in as `argv`.
+
+* When auth flag is 0xFE, the `dynamic linking` method is used. The preimage's memory layout will be as follows:
+```
+dynamic library code hash (32 bytes)
+dynamic library hash type (1 byte)
+pubkey blake160 hash （20 bytes)
+```
+
+It loads the dynamic linking libraries via `code hash` and `hash type`, and gets the entry function named 
+`validate_signature`. The entry function is expected to have following C API:
+```C
+int validate_signature(void *prefilled_data, const uint8_t *signature_buffer,
+    size_t signature_size, const uint8_t *message_buffer, size_t message_size,
+    uint8_t *pubkey_hash, size_t *pubkey_hash_len);
+```
+Then the entry function is called to validate the message and signature. The `pubkey_hash` returned from the entry
```

Comment:
same question, what does the message here mean specifically?

Reply:
The description of the `message` is added in the section `Omnilock Witness`.